assets/src/admin/paired-browsing/app.js

```diff
@@ -0,0 +1,294 @@
+	 */
+	removeAmpQueryVars( url ) {
+		return removeQueryArgs( url, ampSlug, ampPairedBrowsingQueryVar, ampValidationErrorsQueryVar );
+	}
+
+	/**
+	 * Adds the AMP query variable to the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL with the AMP query variable.
+	 */
+	addAmpQueryVar( url ) {
+		return addQueryArgs(
+			url,
+			{
+				[ ampSlug ]: '',
+			},
+		);
+	}
+
+	/**
+	 * Adds the AMP paired browsing query variable to the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL with the AMP paired browsing query variable.
+	 */
+	addPairedBrowsingQueryVar( url ) {
+		return addQueryArgs(
+			url,
+			{
+				[ ampPairedBrowsingQueryVar ]: '1',
+			},
+		);
+	}
+
+	/**
+	 * Removes the URL hash from the supplied URL.
+	 *
+	 * @param {string} url URL string.
+	 * @return {string} Modified URL without the hash.
+	 */
+	removeUrlHash( url ) {
+		const parsedUrl = new URL( url );
+		parsedUrl.hash = '';
+		return parsedUrl.href;
+	}
+
+	/**
+	 * Checks if a URL has the 'amp_validation_errors' query variable.
+	 *
+	 * @param {string} url URL string.
+	 * @return {boolean} True if such query var exists, false if not.
+	 */
+	urlHasValidationErrorQueryVar( url ) {
+		return hasQueryArg( url, 'amp_validation_errors' );
+	}
+
+	/**
+	 * Registers the provided client window with its parent, so that it can be managed by it.
+	 *
+	 * @param {Window} win Document window.
+	 */
+	registerClientWindow( win ) {
+		let oppositeWindow;
+
+		if ( win === this.ampIframe.contentWindow ) {
+			if ( ! this.documentIsAmp( win.document ) ) {
+				if ( this.urlHasValidationErrorQueryVar( win.location.href ) ) {
+					/*
+					 * If the AMP page has validation errors, mark the page as invalid so that the
+					 * 'disconnected' overlay can be shown.
+					 */
+					this.ampPageHasErrors = true;
+					this.toggleDisconnectOverlay( this.ampIframe );
+					return;
+				} else if ( win.document.querySelector( 'head > link[rel=amphtml]' ) ) {
+					// Force the AMP iframe to always have an AMP URL, if an AMP version is available.
+					win.location.replace( this.addAmpQueryVar( win.location.href ) );
+					return;
+				}
+
+				/*
+				 * If the AMP iframe has loaded a non-AMP page and none of the conditions above are
+				 * true, then explicitly mark it as having errors and display the 'disconnected
+				 * overlay.
+				 */
+				this.ampPageHasErrors = true;
+				this.toggleDisconnectOverlay( this.ampIframe );
+				return;
+			}
+
+			this.ampPageHasErrors = false;
+			oppositeWindow = this.nonAmpIframe.contentWindow;
+		} else {
+			// Force the non-AMP iframe to always have a non-AMP URL.
+			if ( this.documentIsAmp( win.document ) ) {
+				win.location.replace( this.removeAmpQueryVars( win.location.href ) );
+				return;
+			}
+
+			// Update the link used for exiting paired browsing.
+			this.exitLink.href = this.nonAmpIframe.contentWindow.location.href;
```

Comment:
As opposed to being the non-AMP window's URL, should this not rather persist whether the user was on an AMP or non-AMP page when they started paired browsing in the first place? If I was on an AMP page on the frontend, and I click “Start paired browsing” then when I exit paired browsing, then I should be put back onto an AMP page on the frontend. Right?